rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -381,6 +386,65 @@ std::string Recorder::serialized_offered_qos_profiles_for_topic(const std::strin
   return YAML::Dump(offered_qos_profiles);
 }
 
+std::string type_hash_to_string(const rosidl_type_hash_t & type_hash)
+{
+  if (type_hash.version == 0) {
+    // version is unset, this is an empty type hash.
+    return "";
+  }
+  if (type_hash.version > 1) {
+    // this is a version we don't know how to serialize
+    ROSBAG2_TRANSPORT_LOG_WARN_STREAM(
+      "attempted to stringify type hash with unknown version " << type_hash.version);
+    return "";
+  }
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  char * stringified_type_hash = nullptr;
+  rcutils_ret_t status = rosidl_stringify_type_hash(&type_hash, allocator, &stringified_type_hash);
+  std::string result = "";
+  if (status == RCUTILS_RET_OK) {
+    result = stringified_type_hash;
+  }
+  if (stringified_type_hash != nullptr) {
+    allocator.deallocate(stringified_type_hash, allocator.state);
+  }
+  return result;
+}
+
+std::string Recorder::type_description_hash_for_topic(const std::string & topic_name) const
+{
+  auto endpoints = this->get_publishers_info_by_topic(topic_name);
+  rosidl_type_hash_t result_hash = rosidl_get_zero_initialized_type_hash();
+  for (const auto & info : endpoints) {
+    // If all endpoint infos provide the same type hash, return it. Otherwise return an empty
+    // string to signal that the type description hash for this topic cannot be determined.
+    const auto endpoint_hash = info.topic_type_hash();
+    if (endpoint_hash.version == 0) {
+      continue;
+    }
+    if (result_hash.version == 0) {
+      result_hash = endpoint_hash;
+      continue;
+    }
+    bool difference_detected = (endpoint_hash.version != result_hash.version);
+    for (int i = 0; i < ROSIDL_TYPE_HASH_SIZE; ++i) {
+      if (result_hash.value[i] != endpoint_hash.value[i]) {
+        difference_detected = true;
+      }
+    }
+    if (difference_detected) {
```

Comment:
```suggestion
if (0 != memcmp(endpoint_hash, result_hash, sizeof(rosidl_type_hash_t))) {
```

Reply:
I don't want to compare padding bytes if I can help it - could be reduced down to two comparisons though.

Reply:
I'm sorry, I don't know what you're talking about:

```
#include <stdio.h>
#include <stdint.h>

typedef struct rosidl_type_hash_s
{
  uint8_t version;
  uint8_t value[32];
} rosidl_type_hash_t;

int main()
{
  printf("Hash is %zu bytes long\n", sizeof(rosidl_type_hash_t));
  return 0;
}
```

Output:

```
3:07:47 ➜  scratch gcc hash.c 
3:07:50 ➜  scratch ./a.out                                                                                               
Hash is 33 bytes long
```

Reply:
Maybe the compiler can unroll your loop but I doubt you're actually reducing instruction count with this more verbose construction...

Reply:
Or if you are then, are you reducing loads which are going to be the majority of the time spent here? I'd optimize your development time for skipping the full `vector` copy that's happening in the `get_publishers_info` call.

Reply:
Or am I missing something, and `memcmp` isn't reliable for struct comparison? Open to counterexample

Reply:
Looking further into it, it seems that generally speaking compilers won't add any padding to this particular struct shape, `uint8_t, uint8_t[32]`, since all the members are the same size. But they're not guaranteed not to. Should have added an equality function to `rosidl_runtime_c`. So even if it works yeah it's not guaranteed reliable, maybe someday we change something to have non-byte values.

For now this should be guaranteed correct and a little more compact?
```
if (a.version == b.version && !memcmp(a.value, b.value, sizeof(a.value))
```

Reply:
@emersonknapp default structure packaging alignment could be redefined at least for GCC
>  #pragma pack(push[,n]) pushes the current alignment setting on an internal stack and then optionally sets the new alignment.
> #pragma pack(pop) restores the alignment setting to the one saved at the top of the internal stack (and removes that stack entry). Note that #pragma pack([n]) does not influence this internal stack; thus it is possible to have #pragma pack(push) followed by multiple #pragma pack(n)

https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gcc/Structure_002dPacking-Pragmas.html

I wouldn't use memcmp(..) alongside with sizeof(struct).  in embedded programming this is a common mistakes and pitfal.